# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x + 2
print(add_two)
add_two

<function add_two at 0x111bee200>


<function __main__.add_two(x)>

In [3]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [4]:
print(lambda x: x + 2)

<function <lambda> at 0x111beede0>


In [5]:
add_two(10)

12

In [6]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)
(lambda x: x + 2)(10)

12

In [8]:
num = 25

num_plus_two = (lambda x: x + 2)(num)
print(num_plus_two)

27


#### Saving to a Variable

In [12]:
named_lambda = lambda y: y**2

print(named_lambda)
print(type(named_lambda))

<function <lambda> at 0x111bee480>
<class 'function'>


In [11]:
named_lambda(10)

100

In [13]:
def my_function(y):
    return y ** 2 

another_function = my_function 

print(another_function)

<function my_function at 0x111bef1a0>


In [14]:
another_function(6)

36

#### Multiple Inputs

In [15]:
def multiply(x, y, z): 
    return x * y * z

multiply(5, 6, 7)

210

In [16]:
multiply(10, 11, 12)

1320

In [17]:
# Syntax: lambda param1, param2, etc: expression
lambda x, y, z: x * y * z

<function __main__.<lambda>(x, y, z)>

In [18]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)
(lambda x, y, z: x * y * z)(3, 4, 5)

60

In [19]:
def accept_args(*args):
    return args

accept_args('a', 'b', 'c', 'd', 'e')

('a', 'b', 'c', 'd', 'e')

In [22]:
lambda *args: args

<function __main__.<lambda>(*args)>

In [31]:
(lambda *args:args) (1, 2, 3)

(1, 2, 3)

In [26]:
def accept_kwargs(**kwargs):
    return kwargs

accept_kwargs(name='Kadeeja', age=28, city='Chicago')

{'name': 'Kadeeja', 'age': 28, 'city': 'Chicago'}

In [27]:
(lambda **kwargs: kwargs)(name='Kadeeja', age=28, city='Chicago')

{'name': 'Kadeeja', 'age': 28, 'city': 'Chicago'}

In [34]:
def get_evens(nums):
    output = []
    for num in nums:
        if num % 2 == 0:
            output.append(num)
    return output
get_evens([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

[2, 4, 6, 8, 10]

In [33]:
(lambda nums: [num for num in nums if num % 2 == 0])([5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

[10, 20, 30, 40, 50]

#### Passing a Function into a Function

In [47]:
def apply_to_each(a_list, f):
    
    """
     print the return value of function f appliped to every elemnet in a_list
     a_list -> list of element to execute function f on 
     f -> function to execute
     """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [48]:
def get_first_letter(string):
    return string[0]

In [49]:
months = ['January', 'February', 'March', 'April', 'May', 'June']

In [50]:
apply_to_each(months, get_first_letter)

J
F
M
A
M
J


In [53]:
def reverse_string(string):
    return string [::-1]

In [54]:
apply_to_each(months, reverse_string)

yraunaJ
yraurbeF
hcraM
lirpA
yaM
enuJ


#### Passing a Lambda into a Function

In [55]:
# Call the apply_to_each function, pass in a lambda function as f argument

apply_to_each(months, lambda month: month.upper())

JANUARY
FEBRUARY
MARCH
APRIL
MAY
JUNE


In [56]:
apply_to_each([1, 2, 3, 4, 5], lambda x: x**2)

1
4
9
16
25


In [58]:
apply_to_each(months, lambda s: s[::-1])

yraunaJ
yraurbeF
hcraM
lirpA
yaM
enuJ


In [59]:
apply_to_each(months, len)

7
8
5
5
3
4


#### Returning a Function from a Function

In [63]:
def my_outer_function(name):
    place = 'Coding Temple'
    print(f'Hello {name}, welcome to {place}')
    return place
    
x = my_outer_function('Kadeeja')



Hello Kadeeja, welcome to Coding Temple


In [64]:
print(place) # cant access because place is inside function

NameError: name 'place' is not defined

In [65]:
print(x)

Coding Temple


In [71]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    #Define a function - will be accepted to my_outer_function
    def my_inner_function():
        city = 'Chicago'
        print(f'This is the inner function {name} from {city}')
        
    # Execute the inner function that I have defined
    my_inner_function()
    
    print(f'Goodbye {name} from {city}') #NameError: name 'city' is not defined 
    
my_outer_function

<function __main__.my_outer_function(name)>

In [73]:
my_outer_function('kadeeja') 

Hello kadeeja
This is the inner function kadeeja from Chicago


NameError: name 'city' is not defined

In [75]:
my_inner_function

NameError: name 'my_inner_function' is not defined

In [81]:
def my_outer_function2(name):
    print(f'Hello {name}')
    
    #define a function - will be scoped to my_outer_function
    def my_inner_function2():
        city = 'chicago'
        print(f'This is the inner function {name} from {city}')
        
        
    #return the inner function
    return my_inner_function2

In [82]:
y = my_outer_function2('kadeeja')

Hello kadeeja


In [83]:
print(y)

<function my_outer_function2.<locals>.my_inner_function2 at 0x1128e6340>


In [84]:
y()

This is the inner function kadeeja from chicago


In [85]:
def create_multiplier(num):
    def multiply(x):
        return num * x
    return multiply

In [86]:
five_times = create_multiplier(5)
print(five_times)

<function create_multiplier.<locals>.multiply at 0x1128e7ec0>


In [90]:
print(five_times(10))
print(five_times(5))
print(five_times(15))
print(five_times(2))

50
25
75
10


In [92]:
three_times = create_multiplier(3)
print(three_times)

<function create_multiplier.<locals>.multiply at 0x1129a0720>


In [94]:
print(three_times(3))
print(three_times(5))
print(three_times(7))
print(three_times(9))

9
15
21
27


#### Returning a Lambda from a Function

In [95]:
def new_multiplier(num):
    return lambda x: num * x

In [96]:
four_times = new_multiplier(4)
print(four_times)

<function new_multiplier.<locals>.<lambda> at 0x1129a0ea0>


In [97]:
four_times(10)

40

#### If Statements within Lambdas

In [106]:
def double_or_square(num):
    #if it is even
    if num % 2 == 0:
        # Square it 
        return num ** 2
    # if it is odd
    else:
        #double it 
        return num ** 2
    
print(double_or_square(5))
print(double_or_square(6))
    

25
36


In [108]:
# Ternary: value_if_true if condition else value_if_false
def double_or_square2(num):
    return num ** 2 if num % 2 == 0 else num * 2

print(double_or_square(5))
print(double_or_square(6))

25
36


In [110]:
# As an IIFE
(lambda num: num **2 if num % 2 == 0 else num*2)(6)

36

In [113]:
apply_to_each([1, 2, 3, 4, 5, 6, 7 , 8 ,9, 10], lambda num: num **2 if num % 2 == 0 else num * 2)

2
4
6
16
10
36
14
64
18
100


#### In-Class Exercise #1 <br>
<p>Write a lambda function that takes in three arguments <code>a</code>,<code>b</code>,<code>c</code> and returns the product of <code>a</code> and the difference of <code>b</code> and <code>c</code></p>

In [ ]:
# Example 1 
# a = 10, b = 5, c = 2
# Output: 30

# Example 2
# a = 20, b = 10, c = 5
# Output: 100

In [117]:
(lambda a, b, c: a * (b-c))(20,10,5)





100

In [118]:
(lambda a, b, c: a * (b-c))(10,5,2)

30

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [119]:
help(map)


Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [130]:
def apply_to_each(a_list, f):
    
    """
     print the return value of function f appliped to every elemnet in a_list
     a_list -> list of element to execute function f on 
     f -> function to execute
     """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [131]:
apply_to_each(months, lambda m: m[::-1])

yraunaJ
yraurbeF
hcraM
lirpA
yaM
enuJ


In [132]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)
months2 = ['July', 'August', 'September', 'October', 'November', 'December']

def reverse_title_case(string):
    return string[0].lower() + string[1:].upper

In [133]:
reverse_title_case('example')

TypeError: can only concatenate str (not "builtin_function_or_method") to str

In [122]:
list(map(reverse_title_case, months2))

TypeError: can only concatenate str (not "builtin_function_or_method") to str

In [126]:
map_return = map(reverse_title_case, months2)
print(map_return)

In [127]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [129]:
next(map_return)

In [140]:
map_return = map(reverse_title_case, months2)

for yielded_val in map_return:
    print(yielded_val)

TypeError: can only concatenate str (not "builtin_function_or_method") to str

In [135]:
x_values = [1, 2, 3, 4, 5]
y_values = [10, 20, 30, 40, 50]

def multiply(x, y):
    print(f"x={x}, y={y}")
    return x*y

In [137]:
# the number of positional arguments required for the function should be equal to the number of iterables
# e.g ultiply takes in teo areguments, so we pass in two iterables (x_values, y_values)

list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50


[10, 40, 90, 160, 250]

In [144]:
x_values2 = [1, 2, 3, 4, 5]
y_values = [10, 20, 30, 40, 50]


list(map(multiply, x_values, y_values))


x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50


[10, 40, 90, 160, 250]

In [148]:
x_values2 = [1, 2, 3, 4, 5]
y_values = [10, 20, 30, 40, 50]
nums_multiply = []

for i in range(len(x_values)):
    return_val = multiply(x_values[i], y_values[i])
    nums_multiply.append(return_val)
    
print(nums_multiply)

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
[10, 40, 90, 160, 250]


In [149]:
def return_hello_world(val):
    return'Hello World'

list(map(return_hello_world, months))

['Hello World',
 'Hello World',
 'Hello World',
 'Hello World',
 'Hello World',
 'Hello World']

#### Using Lambda's with Map

In [150]:
nums = [2, 3, 54, 7, 8, 3, 2, 6, 45, 3]

list(map(lambda num: num**2, nums))

[4, 9, 2916, 49, 64, 9, 4, 36, 2025, 9]

In [151]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'

unicode = list(map(lambda letter: ord(letter), alphabet))
print(unicode)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122]


In [152]:
vowels = ('a', 'e', 'i', 'o', 'u')

upper_vowels = tuple(map(lambda v: v.upper(), vowels))

print(upper_vowels)

('A', 'E', 'I', 'O', 'U')


In [153]:
list(range(0,101,10))

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]

In [157]:
list(map(lambda name: 'Hello ' + name,  ['Brian', 'Sarah', 'Kevin', 'Mike']))

['Hello Brian', 'Hello Sarah', 'Hello Kevin', 'Hello Mike']

In [158]:
for value in map(lambda name:'Hello ' + name,  ['Brian', 'Sarah', 'Kevin', 'Mike']):
    print(value)


Hello Brian
Hello Sarah
Hello Kevin
Hello Mike


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if it is odd</p>

In [164]:
numbers = range(1,11)

# [0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]
result = list(map(lambda x: x * 2 if x % 2 == 0 else x / 2, numbers))
print(result)

def double_even_half_odd(number):
    if number % 2 == 0:
        return number * 2
    else:
        return number / 2

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


In [166]:
result2 = list(map(lambda num: num*2 if num%2==0 else num/2, numbers))
print(result2)

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


In [168]:
cubed = list(map(lambda x: x**3, numbers))
print(cubed)

[1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]


In [ ]:
y = 10 

cubed

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [169]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [173]:
names = ['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron', 'Ben']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(map(starts_with_a, names))
print(a_names_map)

[False, False, True, False, True, True, False, False, False]


In [174]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['Alex', 'Amy', 'Andy']


In [175]:
nums = [num for num in range(1, 11)]
print(nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [178]:
def remainder(num):
    #Will return the remainder of num after being divided by 3
    return num % 3

In [179]:
nums_map = list(map(remainder, nums))
print(nums_map)

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]


In [182]:
nums_filter = list(filter(remainder, nums))
print(nums_filter)

[1, 2, 4, 5, 7, 8, 10]


In [183]:
for i in range(-5, 6):
    print(i, bool(i))

-5 True
-4 True
-3 True
-2 True
-1 True
0 False
1 True
2 True
3 True
4 True
5 True


#### Using Lambda's with Filter()

In [186]:
print(names)

['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron', 'Ben']


In [190]:
b_names = list(filter(lambda name: name[0].upper() == 'B', names))
print(b_names)

['Bob', 'Billy', 'Barbara', 'Ben']


In [192]:
sentence = 'I LIVE AT 123 REAL STREET, CITY, STATE 8888, MY NUMBER IS 555 55 55'

digits = list(filter(lambda char: char.isdigit(), sentence))
print(digits)


['1', '2', '3', '8', '8', '8', '8', '5', '5', '5', '5', '5', '5', '5']


#### In-Class Exercise #3 <br>
<p>Use the filter function to that take in a list of cities as strings and returns a list of cities with at least two words in the name</p>

In [195]:
cities = ['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']
# Output: ['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge'] 

print(cities)

['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']


In [222]:
two_names = list(filter(lambda c: len(c.split())>=2, cities))
print(two_names)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


In [209]:
def two_names(city):
    for cit in city:
        if cit == ' ':
            return city 


two_name_cities = list(filter(two_names, cities))
print(two_name_cities)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


In [224]:
def multi_word_city(city):
    return len(city.split()) >= 2

multi_word_cities = list(filter(multi_word_city, cities))
print(multi_word_cities)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

In [ ]:
import this

In [225]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



#### Syntax

In [226]:
def add_nums(x, y):
    print('='*28)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x+y}")
    return x+y

In [228]:
result_add = reduce(add_nums, [1, 2, 3, 4, 5])
print(result_add)

x: 1
y: 2
returns: 3
x: 3
y: 3
returns: 6
x: 6
y: 4
returns: 10
x: 10
y: 5
returns: 15
15


In [239]:
total = 0 
for num in [1, 2, 3, 4, 5]:
    print(f)

NameError: name 'f' is not defined

In [230]:
def find_max(a, b):
    print('='*20)
    print(f"a: {a}")
    print(f"b: {b}")
    print(f"returns: {a if a > b else b}")
    return a if a > b else b

In [233]:
from random import randint
nums = [randint(1,50) for i in range(10)]
print(nums)

[18, 40, 23, 34, 42, 36, 34, 1, 28, 8]


In [232]:
max_val = reduce(find_max, nums)
print(max_val)

a: 32
b: 30
returns: 32
a: 32
b: 12
returns: 32
a: 32
b: 29
returns: 32
a: 32
b: 32
returns: 32
a: 32
b: 37
returns: 37
a: 37
b: 9
returns: 37
a: 37
b: 24
returns: 37
a: 37
b: 47
returns: 47
a: 47
b: 16
returns: 47
47


In [234]:
max_value = nums[0]
for num in nums:
    if num > max_value:
        max_value = num
print(max_value)

42


#### Using Lambda's with Reduce()

In [235]:
print(nums)

[18, 40, 23, 34, 42, 36, 34, 1, 28, 8]


In [237]:
total = reduce(lambda a,b: a if b>b else b, nums)
print(max_val)

47


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [246]:
nums = [2, 4, 6, 8, 10]
# Output: 3840
#result_add = reduce(add_nums, [1, 2, 3, 4, 5])
#print(result_add)

product = reduce(lambda x, y: x*y, nums)
print(product)







3840


In [247]:
def multiply_nums(x, y):
    #print('='*2)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x*y}")
    return x*y

In [244]:
#result_add = reduce(add_nums, [1, 2, 3, 4, 5])
#print(result_add)

result_multiply = reduce(multiply_nums, [2, 4, 6, 8, 10])
print(result_multiply)

==
x: 2
y: 4
returns: 8
==
x: 8
y: 6
returns: 48
==
x: 48
y: 8
returns: 384
==
x: 384
y: 10
returns: 3840
3840


6483626311680


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [ ]:
! <- FACTORIAL 
x! = x *(x-1)!
5! = 5 * 4!
4! = 4 *3!
3! = 3 * 2!

In [248]:
def capitalize(name):
    # Takes in a name, upper case the first letter and lower case the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize the first name, take in a last name, capitalize the last name and 
    # then returns the concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)


def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
greet_person('bRIaN', 'stANtOn', 'tUEsdaY')

Hello Brian Stanton, it sure is beautiful on this Tuesday


In [249]:
import time

def func_one():
    print("Function One Started")
    time.sleep(5)
    print("Function One Ended")
    
    
def func_two():
    print("Function Two Started")
    func_one()
    print("Function Two Ended")
    
    
func_two()

Function Two Started
Function One Started
Function One Ended
Function Two Ended


#### Implementing a Base Case

In [ ]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

# factorial(x) = x * factorial(x-1)


In [252]:
def factorial(x):
    # set up a base case
    if x <= 1:
        return 1
    else:
        print(f"factorial{x} = {x} * factorial({x-1})")
        return x * factorial(x-1)
factorial(5)

factorial5 = 5 * factorial(4)
factorial4 = 4 * factorial(3)
factorial3 = 3 * factorial(2)
factorial2 = 2 * factorial(1)


120

In [253]:
def factorial(x):
    i = 1
    while i < x:
        output *= i
        i += 1
    return output

factorial(5)

UnboundLocalError: cannot access local variable 'output' where it is not associated with a value

In [257]:
# Can also be used as a way of looping
def loop(a_list):
    print(a_list[0])
    loop(a_list[1:])

loop(['red', 'orange', 'yellow', 'green', 'blue', 'indigo', 'violet'])

red
orange
yellow
green
blue
indigo
violet


IndexError: list index out of range

In [258]:
def is_even(number):
    if number == 1:
        return False
    elif number == 0:
        return true
    else:
        return is_even(number - 2)
is_even(21)

False

#### Writing a Factorial Function

In [ ]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

# factorial(x) = x * factorial(x-1)



In [ ]:
# NEED SOME SORT OF BASE CASE

# def factorial(num):
#     return num * factorial(num - 1)

# factorial(3)

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [259]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC", "     "]

print(places)

[' ', 'Argentina', ' ', 'San Diego', '', '  ', '', 'Boston', 'New York', 'DC', '     ']


In [297]:
filtered_places = []
for place in places:
    if place.strip():
        filtered_places.append(place)
    

print(filtered_places) 

['Argentina', 'San Diego', 'Boston', 'New York', 'DC']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name *case insensitive*...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [321]:
authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]

def sort_auth(authors):
    authors.sort(key=lambda x: x.split()[-1].lower())
    
    return authors


print(authors)
print(sort_auth(authors))



['Joel Carter', 'Victor aNisimov', 'Andrew P. Garfield', 'David hassELHOFF', 'Gary A.J. Bernstein']
['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [345]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]

convert_to_far = list(map(lambda x: (x[0], (9/5)* x[1] + 32), places))

print(convert_to_far)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p> <p>More information on the Fibonacci Sequence <a href="https://en.wikipedia.org/wiki/Fibonacci_sequence" target="_blank">here</a>. Start the sequence with `0,1,1,...`.</p>



In [346]:
def fib(n):
    if (n == 0):
        return 0
    elif (n == 1):
        return 1
    else: return fib(n-2) + fib(n-1)
    
print(fib(8))

21


### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared down to 0, then call the generator and iterate through and print the result.</p>

In [ ]:
# for i in generator(10):
#     print(i)

# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0